src/subtyping.md

```diff
@@ -1,146 +1,148 @@
 # Subtyping and Variance
 
-Subtyping is a relationship between types that allows statically typed
-languages to be a bit more flexible and permissive.
+Rust uses lifetimes to track the relationships between borrows and ownership.
+However, a naive implementation of lifetimes would be either too restrictive,
+or permit undefined behaviour.
 
-Subtyping in Rust is a bit different from subtyping in other languages. This
-makes it harder to give simple examples, which is a problem since subtyping,
-and especially variance, is already hard to understand properly. As in,
-even compiler writers mess it up all the time.
+In order to allow flexible usage of lifetimes
+while also preventing their misuse, Rust uses **subtyping** and **variance**.
 
-To keep things simple, this section will consider a small extension to the
-Rust language that adds a new and simpler subtyping relationship. After
-establishing concepts and issues under this simpler system,
-we will then relate it back to how subtyping actually occurs in Rust.
+Let's start with an example.
 
-So here's our simple extension, *Objective Rust*, featuring three new types:
-
-```rust
-trait Animal {
-    fn snuggle(&self);
-    fn eat(&mut self);
+```rust,ignore
+fn debug<'a>(a: &'a str, b: &'a str) {
+    println!("a = {:?} b = {:?}", a, b);
 }
 
-trait Cat: Animal {
-    fn meow(&self);
+fn main() {
+    let hello: &'static str = "hello";
+    {
+        let world = String::from("world");
+        let world = &world; // 'world has a shorter lifetime than 'static
+        debug(hello, world);
+    }
 }
+```
 
-trait Dog: Animal {
-    fn bark(&self);
-}
+In a conservative implementation of lifetimes, since `hello` and `world` have differing lifetimes,
+we might see the following error:
+
+```text
+error[E0308]: mismatched types
+ --> src/main.rs:10:16
+   |
+10 |         debug(hello, world);
+   |                      ^
+   |                      |
+   |                      expected `&'static str`, found struct `&'world str`
 ```
 
-But unlike normal traits, we can use them as concrete and sized types, just like structs.
+This would be rather unfortunate. In this case,
+what we want is to accept any type that lives *at least as long* as `'world`.
+Let's try using subtyping with our lifetimes.
 
-Now, say we have a very simple function that takes an Animal, like this:
+## Subtyping
 
-<!-- ignore: simplified code -->
-```rust,ignore
-fn love(pet: Animal) {
-    pet.snuggle();
-}
-```
+Subtyping is the idea that one type can be used in place of another.
 
-By default, static types must match *exactly* for a program to compile. As such,
-this code won't compile:
+Let's define that `Sub` is a subtype of `Super` (we'll be using the notation `Sub: Super` throughout this chapter).
```

Comment:
Just to make sure this doesn't get lost: I don't think we should use this notation. `Sub <: Super` is correct. `Sub: Super` is too confusing with outlives relationship, and generally the opposite of what is (potentially) expected.

Reply:
Yeah these syntax suggestions are growing on me. I think i agree with them now and I will perform the changes

Reply:
Why not use existing symbols like `⊂` for this, rather than inventing our own?

Reply:
`<:` is well-used to indicate subtyping and is easy to type.

Reply:
I think that `<:` seems to be too widespread/common in this scenario not to use it. That being said, when generic types are involved, things start to be quite noise, angle-bracket-wise:

```console
Result<Vec<&'static str>> <: Result<Vec<&'_ str>>
```

So `⊂:`, besides being a nicer smiling face 😄, could be a decent middle-ground with @mejrs's proposal:

```console
Result<Vec<&'static str>> ⊂: Result<Vec<&'_ str>>
```

But, it would have the sad side-effect that it wouldn't be that nice in conjunction with `⊇` for regions, and I _really_ find `⊇` for regions to be an incredibly useful notation:

```rust
for<'lt>
      // 'static ⊇ 'lt, and thus:
    &'static str ⊂ &'lt str
,
```

So, if we were not to use `<:`, I'd rather we use a notation with _motion_ to represent subtyping:

```rust
for<'lt>
      // 'static ⊇ 'lt, and thus:
    &'static str → &'lt str
,
Result<Vec<&'static str>> → Result<Vec<&'_ str>>
```

Something like this is, imho, the most readable notation for people unacquainted with type theory (which, let's not forget it, is the target audience for this kind of chapters).